tensorflow/compiler/xla/service/gpu/matmul_utils.cc

```diff
@@ -752,53 +804,112 @@ Status MatmulPlan::DoMatmul(se::Stream* stream, se::DeviceMemoryBase a_buffer,
+      profile_result);
 }
 
 Status MatmulPlan::ExecuteOnStream(
     se::Stream* stream, se::DeviceMemoryBase a_buffer,
     se::DeviceMemoryBase b_buffer, se::DeviceMemoryBase c_buffer,
     se::DeviceMemoryBase d_buffer, se::DeviceMemoryBase bias_buffer,
+    se::DeviceMemoryBase a_scale_buffer, se::DeviceMemoryBase b_scale_buffer,
+    se::DeviceMemoryBase c_scale_buffer, se::DeviceMemoryBase d_scale_buffer,
+    se::DeviceMemoryBase d_amax_buffer,
     const se::cuda::BlasLt::MatmulAlgorithm& algorithm,
     se::ScratchAllocator& scratch_allocator,
     se::blas::ProfileResult* profile_result) const {
   if (must_swap_operands_) {
     std::swap(a_buffer, b_buffer);
   }
 
-  switch (plan_.d_desc.type()) {
-    case CUDA_R_16F:
-      return DoMatmul<Eigen::half, float>(stream, a_buffer, b_buffer, c_buffer,
-                                          d_buffer, bias_buffer, algorithm,
-                                          scratch_allocator, profile_result);
-    case CUDA_R_16BF:
-      return DoMatmul<Eigen::bfloat16, float>(
-          stream, a_buffer, b_buffer, c_buffer, d_buffer, bias_buffer,
-          algorithm, scratch_allocator, profile_result);
-    case CUDA_R_32F:
-      return DoMatmul<float>(stream, a_buffer, b_buffer, c_buffer, d_buffer,
-                             bias_buffer, algorithm, scratch_allocator,
-                             profile_result);
-    case CUDA_R_64F:
-      return DoMatmul<double>(stream, a_buffer, b_buffer, c_buffer, d_buffer,
-                              bias_buffer, algorithm, scratch_allocator,
-                              profile_result);
-    case CUDA_C_32F:
-      return DoMatmul<complex64>(stream, a_buffer, b_buffer, c_buffer, d_buffer,
-                                 bias_buffer, algorithm, scratch_allocator,
-                                 profile_result);
-    case CUDA_C_64F:
-      return DoMatmul<complex128>(stream, a_buffer, b_buffer, c_buffer,
-                                  d_buffer, bias_buffer, algorithm,
-                                  scratch_allocator, profile_result);
-    default:
-      return InternalError("Unexpected dtype");
+  std::tuple<cudaDataType_t, cudaDataType_t, cudaDataType_t, cudaDataType_t>
+      operand_types{plan_.a_desc.type(), plan_.b_desc.type(),
+                    plan_.c_desc.type(), plan_.d_desc.type()};
+
+#define TYPED_MATMUL_F8(ATYPE, BTYPE, CTYPE, DTYPE)                         \
+  if (operand_types == std::make_tuple(ATYPE, BTYPE, CTYPE, DTYPE)) {       \
+    return DoMatmul<float, CudaToNativeT<ATYPE>::type,                      \
+                    CudaToNativeT<BTYPE>::type, CudaToNativeT<CTYPE>::type, \
+                    CudaToNativeT<DTYPE>::type>(                            \
+        stream, a_buffer, b_buffer, c_buffer, d_buffer, bias_buffer,        \
+        a_scale_buffer, b_scale_buffer, c_scale_buffer, d_scale_buffer,     \
+        d_amax_buffer, algorithm, scratch_allocator, profile_result);       \
+  }
```

Comment:
I wonder if we need template parameters in DoMatmul at all. DoMatmul here calls the templated [BlasLt::DoMatmul](https://github.com/tensorflow/tensorflow/blob/46daeddbb728ac29146f8c9e3b17db1b27dbf9d0/tensorflow/compiler/xla/stream_executor/cuda/cuda_blas_lt.h#L149), which calls an untemplated overload of [BlasLt::DoMatmul](https://github.com/tensorflow/tensorflow/blob/46daeddbb728ac29146f8c9e3b17db1b27dbf9d0/tensorflow/compiler/xla/stream_executor/cuda/cuda_blas_lt.h#L190). It seems the template parameters are only used to check argument types match the descriptor types. I don't think this is a particularly useful check, since the caller (XLA or a TF OpKernel) will already ensure the correct types are passed. Perhaps we should get rid of the template arguments, which would simply the logic here (you would still need to check the combination of dtypes are correct though).

In any case, no need to change anything for this PR. This can be addressed in a future PR.